misk-hibernate-testing/src/main/kotlin/misk/jdbc/CrossShardQueryDetector.kt

```diff
@@ -0,0 +1,119 @@
+package misk.jdbc
+
+import com.squareup.moshi.Moshi
+import misk.moshi.adapter
+import misk.okio.split
+import net.ttddyy.dsproxy.ExecutionInfo
+import net.ttddyy.dsproxy.QueryInfo
+import net.ttddyy.dsproxy.listener.QueryExecutionListener
+import net.ttddyy.dsproxy.support.ProxyDataSource
+import okhttp3.OkHttpClient
+import okhttp3.Request
+import okio.BufferedSource
+import okio.ByteString.Companion.encodeUtf8
+import java.io.EOFException
+import javax.inject.Singleton
+import javax.sql.DataSource
+
+data class Instruction(
+  val Opcode: String,
+  val Input: Instruction?
+) {
+  val isScatter: Boolean
+    get() {
+      return Opcode == "SelectScatter" || Input?.isScatter ?: false
+    }
+}
+
+data class QueryPlan(
+  val Original: String,
+  val Instructions: Instruction,
+  val ExecCount: Int
+) {
+  val isScatter: Boolean
+    get() = Instructions.isScatter
+}
+
+/**
+ * Throws a [CrossShardQueryException] for scatter queries that doesn't have a lookup vindex.
+ * Note: Current implementation is not thread safe and will not work in production.
```

Comment:
this is an interesting comment . . . a non-threadsafe singleton makes me itchy

Reply:
Yeah @systay gave me shit about this too. I think I'm gonna have to rethink how I do this. Currently this is only used in test of course. Do you mind if I mull this over for a bit and return to this in a follow up PR?